llvm_passes/HipPromoteInts.cpp

```diff
@@ -0,0 +1,115 @@
+#include "HipPromoteInts.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define DEBUG_TYPE "hip-promote-ints"
+
+/**
+ * This pass promotes integer types to the next standard bit width.
+ * During optimization of loops, LLVM generates non-standard integer types
+ * such as i33 or i56
+ * 
+ * __global__ void testWarpCalc(int* debug) {
+    int tid = threadIdx.x;
+    int bid = blockIdx.x;
+    int globalIdx = bid * blockDim.x + tid;
+    
+    // Optimizations on this loop will generate i33 types.
+    int result = 0;
+    for(int i = 0; i < tid + 1; i++) {
+        result += i * globalIdx;
+    }
+    
+    // Store using atomic operation
+    atomicExch(&debug[globalIdx], result);
+}
+ * 
+ * https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/2823
+ */
+
+using namespace llvm;
+
+bool HipPromoteIntsPass::isStandardBitWidth(unsigned BitWidth) {
+  return BitWidth == 1 || BitWidth == 8 || BitWidth == 16 || BitWidth == 32 || BitWidth == 64;
+}
+
+unsigned HipPromoteIntsPass::getPromotedBitWidth(unsigned Original) {
+  if (Original <= 8) return 8;
+  if (Original <= 16) return 16;
+  if (Original <= 32) return 32;
+  return 64;
+}
+
+PreservedAnalyses HipPromoteIntsPass::run(Module &M, ModuleAnalysisManager &AM) {
+  bool Changed = false;
+  
+  for (Function &F : M) {
+    LLVM_DEBUG(dbgs() << "[HipPromoteInts] Analyzing function: " << F.getName() << "\n");
+    
+    for (BasicBlock &BB : F) {
+      // Use a vector to store instructions that need modification
+      std::vector<Instruction*> WorkList;
+      for (Instruction &I : BB) {
+        WorkList.push_back(&I);
+      }
+      
+      // Process the worklist safely outside the BB iteration
+      for (Instruction *I : WorkList) {
+        if (auto *IntTy = dyn_cast<IntegerType>(I->getType())) {
+          if (!isStandardBitWidth(IntTy->getBitWidth())) {
+            LLVM_DEBUG(dbgs() << "[HipPromoteInts] Found non-standard type in result: " << *I << "\n");
+            
+            unsigned NextStdSize = getPromotedBitWidth(IntTy->getBitWidth());
+            Type *PromotedType = Type::getIntNTy(M.getContext(), NextStdSize);
+            
+            LLVM_DEBUG(dbgs() << "[HipPromoteInts] Promoting from i" << IntTy->getBitWidth() 
+                      << " to i" << NextStdSize << "\n");
+            
+            // Update the instruction type
+            I->mutateType(PromotedType);
```

Comment:
"Note that this is an extremely dangerous operation which can create completely invalid IR very easily. It is strongly recommended that you recreate IR objects with the right types instead of mutating them in place." [[source]](https://llvm.org/doxygen/classllvm_1_1Value.html#ac0f09c2c9951158f9eecfaf7068d7b20)

Please, do what the LLVM documentation suggests.